HACKING.adoc

```diff
@@ -30,7 +30,12 @@ make world.opt
 make runtop
 ----
 
-4. Hack frenetically and keep rebuilding.
+4. Hack frenetically and keep rebuilding. You can speed up future builds
+by using `ocamlc.opt` to compile the compiler:
++
+----
+cp ocamlc.opt boot/
+----
```

Comment:
I'm unconvinced by having this in the "intro for beginners" section of HACKING.adoc. Could we move this to the "Development tips and tricks" section, and maybe say a word about the risk of having a stale ocamlc.opt compiler (if you move between branches for example)?

Reply:
What's the risk? The makefile checks whether `ocamlc.opt` is newer than `ocamlc` now.

It seemed appropriate to tell people about this feature at the point where we advise them to "keep rebuilding". Why bury the fact that there's a command that makes "keep rebuilding" take half as long?

Reply:
I think it's correct that the newer test as it stands guards against `ocamlc.opt` being physically unable to build the compiler because of bootstrap issues, but what about if it's compiled with different (experimental) semantics which didn't require a `make promote`/bootstrap?

I agree with @stedolan that this should be here, where it's more likely to be seen and so played with. We can only go so far to make the experience of hacking on the compiler a risk-free endeavour for beginners 🙂

Reply:
I disagree. Today we *can* give small hacking tasks to almost-complete OCaml beginners, and HACKING.adoc has enough information for them to easily get started building things.

`ocamlc.opt` should not be in the introduction for two reasons:

- There is a risk of things going wrong. Not just a miscompilation due to a stale binary; from a beginner perspective "the build errors/warns because of this out-of-date file" can be disturbing as well.
- This section should be as *simple* as possible, avoiding concept overload. Everything that requires a more complex discussion should be held off for later. If we can remove some information from this section, we should.

For newcomers to compiler hacking, I would rather have slower build time than slower understanding of how to get started on the compiler.

Again, the "Development tips and tricks" section is precisely there to give useful tips to people. For example, `make -C testsuite parallel` halves the testsuite running time, and it's there. The `ocamlc.opt` trick should be there as well.

Reply:
I still don't follow point 1 here (out-of-date boot/ocamlc.opt doesn't cause a build error/warning), but point 2 is convincing. I've updated HACKING.adoc.

Reply:
We're both right for point (1): in my mind there *should* be a warning if `boot/ocamlc.opt` is stale (silently ignoring it is not good UI), but you left it out of the perimeter of the present PR -- I'm fine with that. I hope there will be a warning later.